crates/sui-types/src/event.rs

```diff
@@ -1,22 +1,78 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use move_core_types::language_storage::StructTag;
+use move_core_types::language_storage::{ModuleId, StructTag};
+use name_variant::NamedVariant;
 use serde::{Deserialize, Serialize};
 use serde_with::{serde_as, Bytes};
 
-/// User-defined event emitted by executing Move code.
-/// Executing a transaction produces an ordered log of these
+use crate::{
+    base_types::{ObjectID, SequenceNumber, SuiAddress, TransactionDigest},
+    committee::EpochId,
+    messages_checkpoint::CheckpointSequenceNumber,
+};
+
+/// A universal Sui event type encapsulating different types of events
+#[derive(Debug, Clone, PartialEq)]
+pub struct EventEnvelope {
+    /// UTC timestamp in milliseconds since epoch (1/1/1970)
+    timestamp: u64,
+    /// Transaction digest of associated transaction, if any
+    tx_digest: Option<TransactionDigest>,
+    /// Specific event type
+    event: SuiEvent,
+}
+
+impl EventEnvelope {
+    pub fn new(timestamp: u64, tx_digest: Option<TransactionDigest>, event: SuiEvent) -> Self {
+        Self {
+            timestamp,
+            tx_digest,
+            event,
+        }
+    }
+
+    pub fn event_type(&self) -> &'static str {
+        self.event.variant_name()
+    }
+}
+
+/// Specific type of event
 #[serde_as]
-#[derive(Eq, PartialEq, Debug, Clone, Deserialize, Serialize, Hash)]
-pub struct Event {
-    pub type_: StructTag,
-    #[serde_as(as = "Bytes")]
-    pub contents: Vec<u8>,
+#[derive(Eq, Debug, Clone, PartialEq, NamedVariant, Deserialize, Serialize, Hash)]
+pub enum SuiEvent {
+    /// Move-specific event
+    MoveEvent {
+        type_: StructTag,
+        #[serde_as(as = "Bytes")]
+        contents: Vec<u8>,
```

Comment:
I want to add a parsed json object field here since these bytes themselves are not that useful. This does not need to happen in this PR though, I will handle it later

Reply:
Hmmm, yes I would like this as well - basically a deserialized version with fields.  However there are a couple really important details to hash out here, we can chat more offline:
1. JSON does not map well to a number of binary fields or to `vector<u8>` like structures.  
2. `vector<u8>` could be mapped to string or a binary blob, but it should not be mapped to a "json array"

The EventStore could use something like that but JSON would probably not work for the reasons above.  I'm wondering if the deserialized MoveStruct is better.  That could be easily converted to JSON later.

Reply:
Maybe we can store the deserialized MoveStruct for now, this could then easily be converted to JSON, and we also want to use it further for something for EventStore.  WDYT?

Reply:
I think we should deserialise this into move struct as the current RPC API are designed to be universal, BCS bytes might not be very useful for some clients?

Reply:
Converting this should be similar to the `MoveObject` <> `SuiMoveObject` conversion https://github.com/MystenLabs/sui/blob/4ebe1a7aa17a0d0787a0d899c4ec52fc2afe4314/crates/sui-core/src/gateway_types.rs#L246-L279

Reply:
@patrickkuo I added a method to deserialize the event bytes to MoveStruct already.   I think I am leaning towards keeping something deserialized, but I'd like to leave this for subsequent PR.

Reply:
sgtm